compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs

```diff
@@ -1136,7 +1147,7 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         diag: &mut Diag<'_>,
         cause: &ObligationCause<'tcx>,
         secondary_span: Option<(Span, Cow<'static, str>)>,
-        mut values: Option<ValuePairs<'tcx>>,
+        mut values: Option<(ValuePairs<'tcx>, ty::ParamEnv<'tcx>)>,
```

Comment:
i feel like the param env should always be required together with the `cause` :thinking: 

looking at the uses of `note_type_err` I feel like this should be a straightforward change

Reply:
Well, so the param-env of this `ValuePairs` is *conceptually* distinct from param-env of the `TypeError` that may be emitted, since (well, in the future perhaps) the `TypeError` may have instantiated binders in scope as well.

So like, imagine a type error between[^a] `for<T: Foo<Assoc = i32>> fn(<T as Foo>::Assoc)` and `fn(u32)`. The valuepairs will contain the outer types, but the type error will likely be a `TypeError::Sorts` mismatch between `<!T as Foo>::Assoc` (w/ the param-env augmented with `!T: Foo<Assoc = i32>`) and `u32`. I think that *other* param-env may ned to be tracked separately.

I think we could still pass in the "outermost" param-env as a separate, but it seems kinda a waste since that param-env really is tightly associated only w the `ValuePairs`...

 thoughts?

[^a]: Assume we have some sort of higher-ranked fn ptr syntax. This isn't super important, just the binders.

Reply:
I guess you're right :cry: :grin:

vibeck: `ParamEnvAnd` instead of a tuple here?

anyways, r=me